File: android/app/src/main/java/ir/ilam/inspection/util/PersianNumbers.kt
package ir.ilam.inspection.util

/**
 * Digit shaping. Persian digits are for display only; anything that reaches the
 * database, a file name or the sync protocol keeps latin digits.
 */
object PersianNumbers {

    private const val PERSIAN = "۰۱۲۳۴۵۶۷۸۹"
    private const val ARABIC = "٠١٢٣٤٥٦٧٨٩"

    fun toPersian(input: String?): String {
        if (input.isNullOrEmpty()) return ""
        val sb = StringBuilder(input.length)
        for (ch in input) {
            sb.append(if (ch in '0'..'9') PERSIAN[ch - '0'] else ch)
        }
        return sb.toString()
    }

    fun toPersian(value: Long): String = toPersian(value.toString())

    fun toPersian(value: Int): String = toPersian(value.toString())

    /** Trims a trailing `.0` so amperage reads `۶۳` rather than `۶۳٫۰`. */
    fun toPersian(value: Double?): String {
        if (value == null) return ""
        val text = if (value == value.toLong().toDouble()) {
            value.toLong().toString()
        } else {
            value.toString()
        }
        return toPersian(text).replace('.', '٫')
    }

    /** Accepts persian or arabic-indic digits typed by the user. */
    fun toLatin(input: String?): String {
        if (input.isNullOrEmpty()) return ""
        val sb = StringBuilder(input.length)
        for (ch in input) {
            val p = PERSIAN.indexOf(ch)
            val a = ARABIC.indexOf(ch)
            sb.append(
                when {
                    p >= 0 -> '0' + p
                    a >= 0 -> '0' + a
                    ch == '٫' -> '.'
                    else -> ch
                }
            )
        }
        return sb.toString()
    }

    fun parseDoubleOrNull(input: String?): Double? =
        toLatin(input).trim().takeIf { it.isNotEmpty() }?.toDoubleOrNull()

    fun parseIntOrNull(input: String?): Int? =
        toLatin(input).trim().takeIf { it.isNotEmpty() }?.toIntOrNull()

    /** Groups thousands for readable watt totals: `۱۲٬۵۰۰`. */
    fun grouped(value: Double?): String {
        if (value == null) return ""
        val whole = value.toLong()
        val text = whole.toString().reversed().chunked(3).joinToString("٬").reversed()
        return toPersian(text)
    }
}

File: windows/CryptoInspection.Archive/Util/PersianNumbers.cs
using System;
using System.Globalization;
using System.Text;

namespace CryptoInspection.Archive.Util
{
    /// <summary>
    /// Digit shaping. Persian digits are for display and printed output only;
    /// the database, file names and the sync protocol keep latin digits.
    /// </summary>
    public static class PersianNumbers
    {
        private const string Persian = "۰۱۲۳۴۵۶۷۸۹";
        private const string Arabic = "٠١٢٣٤٥٦٧٨٩";

        public static string ToPersian(string input)
        {
            if (string.IsNullOrEmpty(input))
            {
                return string.Empty;
            }

            StringBuilder builder = new StringBuilder(input.Length);
            foreach (char c in input)
            {
                builder.Append(c >= '0' && c <= '9' ? Persian[c - '0'] : c);
            }

            return builder.ToString();
        }

        public static string ToPersian(long value)
        {
            return ToPersian(value.ToString(CultureInfo.InvariantCulture));
        }

        public static string ToPersian(int value)
        {
            return ToPersian(value.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary>Trims a trailing `.0` so amperage reads `۶۳`, not `۶۳٫۰`.</summary>
        public static string ToPersian(double? value)
        {
            if (!value.HasValue)
            {
                return string.Empty;
            }

            return ToPersian(Plain(value)).Replace('.', '٫');
        }

        public static string Plain(double? value)
        {
            if (!value.HasValue)
            {
                return string.Empty;
            }

            double v = value.Value;
            if (Math.Abs(v - Math.Truncate(v)) < 0.0000001)
            {
                return ((long)v).ToString(CultureInfo.InvariantCulture);
            }

            return v.ToString(CultureInfo.InvariantCulture);
        }

        /// <summary>Accepts persian or arabic-indic digits typed by the operator.</summary>
        public static string ToLatin(string input)
        {
            if (string.IsNullOrEmpty(input))
            {
                return string.Empty;
            }

            StringBuilder builder = new StringBuilder(input.Length);
            foreach (char c in input)
            {
                int p = Persian.IndexOf(c);
                int a = Arabic.IndexOf(c);
                if (p >= 0)
                {
                    builder.Append((char)('0' + p));
                }
                else if (a >= 0)
                {
                    builder.Append((char)('0' + a));
                }
                else if (c == '٫')
                {
                    builder.Append('.');
                }
                else
                {
                    builder.Append(c);
                }
            }

            return builder.ToString();
        }

        /// <summary>Grouped thousands for readable watt totals: `۱۲٬۵۰۰`.</summary>
        public static string Grouped(double? value)
        {
            if (!value.HasValue)
            {
                return string.Empty;
            }

            string text = ((long)value.Value).ToString("#,0", CultureInfo.InvariantCulture);
            return ToPersian(text.Replace(",", "٬"));
        }
    }
}
